packages/keyring-controller/src/KeyringController.test.ts

```diff
@@ -3893,20 +4046,16 @@ describe('KeyringController', () => {
       it('should rollback the controller keyrings if the keyring creation fails', async () => {
         const mockAddress = '0x4584d2B4905087A100420AFfCe1b2d73fC69B8E4';
         stubKeyringClassWithAccount(MockKeyring, mockAddress);
-        // Mocking the serialize method to throw an error will
-        // halt the controller everytime it tries to persist the keyring,
-        // making it impossible to update the vault
-        jest
-          .spyOn(MockKeyring.prototype, 'serialize')
-          .mockImplementation(async () => {
-            throw new Error('You will never be able to persist me!');
-          });
+
         await withController(
           { keyringBuilders: [keyringBuilderFactory(MockKeyring)] },
           async ({ controller, initialState }) => {
+            jest.spyOn(controller, 'update' as never).mockImplementation(() => {
+              throw new Error('You will never be able to change me!');
+            });
             await expect(
               controller.addNewKeyring(MockKeyring.type),
-            ).rejects.toThrow('You will never be able to persist me!');
+            ).rejects.toThrow('You will never be able to change me!');
 
```

Comment:
I'm confused, are we mocking `BaseController.update` here?

Reply:
Ok, so it probably needs a comment : ) So basically we want to make sure that all the operations happens before we throw an error to test the rollback. `.update()` is the last thing in the `#updateVault()`, so at this point we can be sure that ie. `keyringsMetadata` has been added, so we can test the rollback correctly. Otherwise we're not really testing rollback, as the `keyringsMetadata` might not be even added.

Reply:
Both `keyring.serialize` and `BaseController.update` should be called during the vault update operation, but generally both would be trying to persist an operation happened _before_ the `this.#updateVault`. 

What we really want to test here is that the rollback operation keeps the controller in a valid state, that is: 

- `this.#keyrings` is still equal to `state.keyrings`
- `this.#keyringsMetadata` is still equal to `state.keyringsMetadata`

Reply:
This to say that the original code should have served the exact same purpose

Reply:
In `#withRollback` function `this.#getSerializedKeyrings` is being called before the `callback()` is triggered, so neither keyring nor keyringsMetada was ever added to the state, because it was throwing an error even before it tried to add new keyring. That's why I had to find another way to test it